consensus/tendermint/core/state_dumper.go

```diff
@@ -0,0 +1,266 @@
+package core
+
+import (
+	"github.com/clearmatics/autonity/common"
+	"github.com/clearmatics/autonity/core/types"
+	"math/big"
+	"time"
+)
+
+type coreStateRequestEvent struct {
+	stateChan chan TendermintState
+}
+
+// VoteState save the prevote or precommit voting status for a specific value.
+type VoteState struct {
+	Value            common.Hash
+	ProposalVerified bool
+	VotePower        uint64
+}
+
+// RoundState save the voting status for a specific round.
+type RoundState struct {
+	Round          int64
+	Proposal       common.Hash
+	PrevoteState   []VoteState
+	PrecommitState []VoteState
+}
+
+// MsgWithHash save the msg and extra field to be marshal to JSON.
+type MsgForDump struct {
+	Message
+	Hash   common.Hash
+	Power  uint64
+	Height *big.Int
+	Round  int64
+}
+
+// TendermintState save an instant status for the tendermint consensus engine.
+type TendermintState struct {
+	// return error code, 0 for okay, -1 for timeout.
+	Code int64
+	// validator address
+	Client common.Address
+
+	// core state of tendermint
+	Height      *big.Int
+	Round       int64
+	Step        uint64
+	Proposal    *common.Hash
+	LockedValue *common.Hash
+	LockedRound int64
+	ValidValue  *common.Hash
+	ValidRound  int64
+
+	// committee state
+	Committee       types.Committee
+	Proposer        common.Address
+	IsProposer      bool
+	QuorumVotePower uint64
+	RoundStates     []RoundState
+	ProposerPolicy  uint64
+
+	// extra state
+	SentProposal          bool
+	SentPrevote           bool
+	SentPrecommit         bool
+	SetValidRoundAndValue bool
+
+	// timer state
+	BlockPeriod           uint64
+	ProposeTimerStarted   bool
+	PrevoteTimerStarted   bool
+	PrecommitTimerStarted bool
+
+	// current height messages.
+	CurHeightMessages []*MsgForDump
+	// backlog msgs
+	BacklogMessages []*MsgForDump
+	// backlog unchecked msgs.
+	UncheckedMsgs []*MsgForDump
+	// Known msg of gossip.
+	KnownMsgHash []common.Hash
+}
+
+func (c *core) CoreState() TendermintState {
+	state := TendermintState{}
+	// send state dump request.
+	var e = coreStateRequestEvent{
+		stateChan: make(chan TendermintState),
+	}
+	go c.sendEvent(e)
+	// wait for response with timeout.
+	timeout := time.After(time.Second)
+	select {
+	case s := <-e.stateChan:
+		state = s
+	case <-timeout:
+		state.Code = -1
+		c.logger.Debug("Waiting for tendermint core state timed out", "elapsed", time.Second)
+	}
+
+	return state
```

Comment:
time out is no longer required

```suggestion
	var e = coreStateRequestEvent{
		stateChan: make(chan TendermintState),
	}
	go c.sendEvent(e)
	return <-e.stateChan
```